openedx/core/djangoapps/content_libraries/tests/test_containers.py

```diff
@@ -580,54 +601,200 @@ def test_publish_container(self) -> None:  # pylint: disable=too-many-statements
-        assert c1_components_after[1]["published_by"] == self.user.username
-        assert c1_components_after[2]["id"] == self.problem_block_2["id"]
-        assert c1_components_after[2]["has_unpublished_changes"] is False
-        assert c1_components_after[2]["published_by"] == self.user.username
-        assert c1_components_after[3]["id"] == self.html_block_2["id"]
-        assert c1_components_after[3]["has_unpublished_changes"] is False
-        assert c1_components_after[3]["published_by"] == self.user.username
-
-        # and container 2 is still unpublished, except for the shared HTML block that is also in container 1:
+        self._verify_publish_state(
+            container_id=self.unit_with_components["id"],
+            expected_childen_len=4,
+            expected_has_unpublished_changes=False,
+            expected_published_by=self.user.username,
+            expected_children_ids=[
+                self.problem_block["id"],
+                self.html_block["id"],
+                self.problem_block_2["id"],
+                self.html_block_2["id"],
+            ],
+        )
+
+        # and unit 2 is still unpublished, except for the shared HTML block that is also in unit 1:
         c2_after = self._get_container(self.unit["id"])
         assert c2_after["has_unpublished_changes"]
         c2_components_after = self._get_container_children(self.unit["id"])
         assert len(c2_components_after) == 2
         assert c2_components_after[0]["id"] == self.html_block["id"]
-        assert c2_components_after[0]["has_unpublished_changes"] is False  # published since it's also in container 1
+        assert c2_components_after[0]["has_unpublished_changes"] is False  # published since it's also in unit 1
         assert c2_components_after[0]["published_by"] == self.user.username
         assert c2_components_after[1]["id"] == html_block_3["id"]
         assert c2_components_after[1]["has_unpublished_changes"]  # unaffected
         assert c2_components_after[1]["published_by"] is None
+
+    def test_publish_subsection(self) -> None:
+        """
+        Test that we can publish the changes to a specific subsection
+        """
+        unit_4 = self._create_container(self.lib["id"], "unit", display_name="Test Unit 4", slug=None)
+
+        # Add units on subsection
+        self._add_container_children(
+            self.subsection["id"],
+            children_ids=[
+                self.unit["id"],
+                unit_4["id"],
+            ]
+        )
+        # Removing the unit with components because the components (children of children) are not published.
+        # If the unit is kept, the subsection continues to have changes even after it is published.
+        self._remove_container_children(
+            self.subsection_with_units["id"],
+            children_ids=[
+                self.unit_with_components["id"],
+            ]
+        )
```

Comment:
@bradenmacdonald @ormsbee Currently, when publishing a container, the children of the children aren't published. In the case of a subsection, the components of a unit that is a child of a subsection aren't published when the subsection is published. This is something that needs to be fixed, right?

Reply:
Yeah, I have a fairly major revamp of https://github.com/openedx/openedx-learning/pull/307 that I want to get into reviewable shape next week. My hope is that the new models I'm introducing (https://github.com/openedx/openedx-learning/issues/317) will make this more straightforward.